### DISCUSSION TITLE
The learnable "gamma" Parameter of BatchNormalization updates when running in inference mode. Is this intended behavior?

### DISCUSSION THREAD

**Discussion by RabJon** [keras-team/keras#21478] [General]
I am aware that the BatchNormalization layer is already much discussed in the Keras docs because of its different behavior during training and inference. However, the fact that the "gamma" parameter of my BatchNormalization layers updated even if I set the "training" argument to False when calling them surprises me a bit. After reading the docs I thought to have understood that it is important to freeze these layers completely when finetuning models and that (for the special case of BatchNormalization Layers) this can be achieved by setting the "training" argument to False when calling the layers. While this special behavior proved true for the "moving_mean" and "moving_variance" parameters of the layers it did not for the "gamma" parameter. Is this intended?

**Comment by Kxrma47** [comment]
Yes, this is intended. `training=False` changes which statistics BatchNormalization uses: it uses `moving_mean` and `moving_variance` and does not update those moving statistics. It does **not** make the trainable affine parameters `gamma` and `beta` non-trainable, so the optimizer may still update them through backpropagation.

To freeze the layer completely, set its `trainable` flag to `False` before compiling the model:

```python
for layer in model.layers:
    if isinstance(layer, keras.layers.BatchNormalization):
        layer.trainable = False

model.compile(optimizer=..., loss=...)
```

If the BatchNormalization layers are inside nested models, recurse through those models as well. Recompile after changing `trainable`, because the compiled training function captures the trainable-variable set.

So the two controls are separate: `training=False` selects inference behavior for the batch statistics, while `layer.trainable=False` prevents `gamma` and `beta` from receiving optimizer updates.
